### DISCUSSION TITLE
how to transform test data to the data which was processed by AutoML when Engineering of features

### DISCUSSION THREAD

**Discussion by feihongloveworld** [h2oai/h2o-3#15872] [Q&A]
now i got a stack model from AutoML，and i‘d like to know how AutoML processed the train data when Engineering of features,
and I want to do the same process on the test data for detail.

**Comment by sebhrusen** [comment]
@feihongloveworld , all transformations done on the features by AutoML—typically just encoding for now, but there's more to come—are also applied automatically on the models you obtain when calling `predict` on your test data, so you don't need to apply those transformations yourself.

**Comment by feihongloveworld** [comment]
> @feihongloveworld , all transformations done on the features by AutoML—typically just encoding for now, but there's more to come—are also applied automatically on the models you obtain when calling `predict` on your test data, so you don't need to apply those transformations yourself.

I knew this point, but I want to know the detail processing of Engineering of features in AutoML, so I could do the same thing on my data for other packages like "shap" in python. if don't do that, same missing features will  prevent the program from running.

**Reply by sebhrusen** [reply]
Each H2O algo may remove features (for example, constant columns, string columns, uids…) and apply a different logic to encode categoricals. Outside of this, H2O AutoML doesn't modify features for now by default (better feature engineering is currently in progress).
As you're using the Py client, right now, you can access the following information from the model you're interested in (not from the `H2OAutoML` object but from the `aml.leader` for example).
```python
aml = H2OAutoML(…)
aml.train(…)
m = aml.leader
# now the list of info you may be interested in
used_cols = m._model_json['output']['names']  # this is an unofficial API but the only way to obtain the list of features that were really used during training, meaning that you can drop the features not listed here if you're trying to reproduce H2O behaviour somewhere else.
enc = m.actual_params['categorical_encoding']  # tells you which algo was used to encode all categorical columns

# note that if the model m is a Stacked Ensemble, then none of this info is available as SE will only calls `predict` for each of the base models, but you can still obtain the exact list of base models used to build the SE
bm = se.base_models
```

I hope this helps. I'm afraid that you can't obtain more details about what models do to the provided features outside what I listed above.
